### PR TITLE
util: replace all std::mutex and std::condition_variable with wrapped absl equivalents.

### PR DESCRIPTION
**Pull request by jmarantz**
*Description*:
>Completely removes dependence on std::mutex and std::condition_variable from envoy test/ and source/.  This enables compile-time enforcement of thread-annotations throughout the codebase, though not all possible annotations have been added yet.

Follow-up PRs will come to add annotations, and to enforce with check_format that std::mutex and std::condition_variable should not re-appear.

*Risk Level*: Low 
*Testing*: //test/...
*Docs Changes*: N/A
*Release Notes*: N/A

### CODE REVIEW

File: source/common/common/thread.h [hunk 1]
```diff
@@ -57,26 +58,40 @@ class MutexBasicLockable : public BasicLockable {
  */
 class CondVar {
 public:
-  // Note that it is not necessary to be holding an associated mutex to call
-  // notifyOne or notifyAll. See the discussion in
-  //     http://en.cppreference.com/w/cpp/thread/condition_variable_any/notify_one
-  // for more details.
+  enum WaitStatus {
```
Comment (ggreenway) [inline source/common/common/thread.h]: enum class, please

File: source/common/common/thread.h [hunk 2]
```diff
@@ -57,26 +58,40 @@ class MutexBasicLockable : public BasicLockable {
...
+   * with thread annotation.
+   */
   void wait(MutexBasicLockable& mutex) noexcept EXCLUSIVE_LOCKS_REQUIRED(mutex) {
     condvar_.Wait(&mutex.mutex_);
   }
   template <class Rep, class Period>
-  void
+
+  /**
+   * @return {WaitStatus} whether the condition timed out or not.
```
Comment (ggreenway) [inline source/common/common/thread.h]: nit: no {} around WaitStatus

File: source/common/common/thread.h [hunk 3]
```diff
@@ -57,26 +58,40 @@ class MutexBasicLockable : public BasicLockable {
  */
 class CondVar {
 public:
-  // Note that it is not necessary to be holding an associated mutex to call
-  // notifyOne or notifyAll. See the discussion in
-  //     http://en.cppreference.com/w/cpp/thread/condition_variable_any/notify_one
-  // for more details.
+  enum WaitStatus {
+    Timeout,
+    NoTimeout,
```
Comment (ggreenway) [inline source/common/common/thread.h]: I think this would be clearer if it were named "Success"

File: source/common/common/thread.h (hunk 1 above)
Comment (jmarantz) [reply source/common/common/thread.h]: I was alright with referencing the status enum values under CondVar, but another indirection is OK too; done.

File: source/common/common/thread.h (hunk 3 above)
Comment (jmarantz) [reply source/common/common/thread.h]: The problem is that this value may also represent a spurious wakeup.  I added a brief comment to that effect. You can only determine success if you know the bool condition for that in control of the application.

File: source/common/common/thread.h (hunk 2 above)
Comment (jmarantz) [reply source/common/common/thread.h]: done

**Pull request merged**